src/rankcorr.jl

```diff
@@ -33,121 +33,204 @@ corspearman(X::RealMatrix) = (Z = mapslices(tiedrank, X, dims=1); cor(Z, Z))
 #
 #######################################
 
-# Knight JASA (1966)
-
-function corkendall!(x::RealVector, y::RealVector)
+# Knight, William R. “A Computer Method for Calculating Kendall's Tau with Ungrouped Data.”
+# Journal of the American Statistical Association, vol. 61, no. 314, 1966, pp. 436–439.
+# JSTOR, www.jstor.org/stable/2282833.
+function corkendall!(x::RealVector, y::RealVector, permx::AbstractVector{<:Integer}=sortperm(x))
     if any(isnan, x) || any(isnan, y) return NaN end
     n = length(x)
     if n != length(y) error("Vectors must have same length") end
 
     # Initial sorting
-    pm = sortperm(y)
-    x[:] = x[pm]
-    y[:] = y[pm]
-    pm[:] = sortperm(x)
-    x[:] = x[pm]
-
-    # Counting ties in x and y
-    iT = 1
-    nT = 0
-    iU = 1
-    nU = 0
-    for i = 2:n
-        if x[i] == x[i-1]
-            iT += 1
-        else
-            nT += iT*(iT - 1)
-            iT = 1
-        end
-        if y[i] == y[i-1]
-            iU += 1
-        else
-            nU += iU*(iU - 1)
-            iU = 1
+    permute!(x, permx)
+    permute!(y, permx)
+    
+    # Use widen to avoid overflows on both 32bit and 64bit
+    npairs = div(widen(n) * (n - 1), 2)
+    ntiesx, ndoubleties, nswaps = widen(0), widen(0), widen(0)
+    k = 0
+
+    @inbounds for i = 2:n
+        if x[i - 1] == x[i]
+            k += 1
+        elseif k > 0
+            # Sort the corresponding chunk of y, so the rows of hcat(x,y) are 
+            # sorted first on x, then (where x values are tied) on y. Hence 
+            # double ties can be counted by calling countties.
+            sort!(view(y, (i - k - 1):(i - 1)))  # Can't use wide integers here
+            ntiesx += div(widen(k) * (k + 1), 2) # Must use wide integers here
+            ndoubleties += countties(y,  i - k - 1, i - 1)
+            k = 0
         end
     end
-    if iT > 1 nT += iT*(iT - 1) end
-    nT = div(nT,2)
-    if iU > 1 nU += iU*(iU - 1) end
-    nU = div(nU,2)
-
-    # Sort y after x
-    y[:] = y[pm]
-
-    # Calculate double ties
-    iV = 1
-    nV = 0
-    jV = 1
-    for i = 2:n
-        if x[i] == x[i-1] && y[i] == y[i-1]
-            iV += 1
-        else
-            nV += iV*(iV - 1)
-            iV = 1
-        end
+    if k > 0
+        sort!(view(y, ((n - k):n)))
+        ntiesx += div(widen(k) * (k + 1), 2)
+        ndoubleties += countties(y,  n - k, n)
     end
-    if iV > 1 nV += iV*(iV - 1) end
-    nV = div(nV,2)
 
-    nD = div(n*(n - 1),2)
-    return (nD - nT - nU + nV - 2swaps!(y)) / (sqrt(nD - nT) * sqrt(nD - nU))
-end
+    nswaps = merge_sort!(y, 1, n)
+    ntiesy = countties(y, 1, n)
 
+    # Calls to float below prevent possible overflow errors when
+    # length(x) exceeds 77_936 (32 bit) or 5_107_605_667 (64 bit)
+    (npairs + ndoubleties - ntiesx - ntiesy - 2 * nswaps) /
+     sqrt(float(npairs - ntiesx) * float(npairs - ntiesy))
```

Comment:
Is overflow still possible given that we use `widen`?

Reply:
Yes, and that fact took me by surprise when 32 bit tests failed without those calls to float.

On 32 bit, `npairs` etc will be of type `Float64`. If `length(x) = 77_937` then

`npairs = 77937 * 77936 / 2 = 3_037_049_016`

so (absent ties) without the calls to `float` we would have the denominator being:

9_223_666_725_586_568_256

which is greater than (2^63)-1 = typemax(Int64) = 

9_223_372_036_854_775_807

On 64 bit the limit on `length(x)` is much greater at 5_107_605_667 (if my calculations are correct).

Reply:
OK. Though on 32-bit `npairs` is `Int64`, not `Float64`, right?

Anyway, to avoid overflow, isn't the simplest approach to do `sqrt(npairs - ntiesx) * sqrt(npairs - ntiesy)`? Or even `... / sqrt(npairs - ntiesx) / sqrt(npairs - ntiesy)`.

Reply:
Ah yes, `Int64`, not `Float64`.

I think  writing `sqrt(float(npairs - ntiesx) * float(npairs - ntiesy))` has the advantage of not letting the result go very slighly outside the range -1 to +1 - wasn't that a minor bug that I fixed quite early on?

Reply:
Not sure. But we can call `clamp` at the end of needed. Are you sure using `float` ensures the result is in the correct range?

Reply:
I'm reasonably sure, yes. If both `x` and `y` have the same ranking then it follows that `nswaps == 0`. If in addition there are no ties then the result (with the current code) becomes equivalent to evaluating:

`npairs/sqrt(npairs * npairs)`

and I would imagine that always evaluates to exactly 1.0 (since `npairs` is an integer)

if there _are_ ties then we have `tiesx == tiesy` and also `ndoubleties == tiesx^2` so the result becomes

`(npairs - tiesx)/sqrt((npairs-tiesx) * (npairs-tiesx))` and once again that will evaluate to exactly 1.0

I think a similar argument applies if `x` and `y` are in exactly opposite order except in that case `nswaps == 2npairs`

So the current code will definitely return exactly 1.0 or -1.0 when it should do so. Not quite the same as proving that the return from the function is always in the range [-1 to 1] I admit.

If we were to change to 

```julia
(npairs + ndoubleties - ntiesx - ntiesy - 2 * nswaps) /
     sqrt(npairs - ntiesx) / sqrt(npairs - ntiesy)

```

then you get
```
julia> corkendall(collect(1:5),collect(1:5))
0.9999999999999999

```
which of course can't be fixed by `clamp`. 

All this makes me think the current code is worth sticking with i.e.

```julia
  (npairs + ndoubleties - ntiesx - ntiesy - 2 * nswaps) /
     sqrt(float(npairs - ntiesx) * float(npairs - ntiesy))

```

Reply:
Thanks for the analysis. I would have thought we could run into problems with values higher than `sqrt(maxintfloat(Float64))` due to rounding, but I can't find an example where it doesn't work.